api/sys_monitor.go

```diff
@@ -0,0 +1,50 @@
+package api
+
+import (
+	"bufio"
+	"context"
+)
+
+// This function returns a channel that outputs strings containing the log messages
+// coming from the server.
+func (c *Sys) Monitor(ctx context.Context, logLevel string) (chan string, error) {
+	r := c.c.NewRequest("GET", "/v1/sys/monitor")
+
+	if logLevel == "" {
+		r.Params.Add("log_level", "INFO")
+	} else {
+		r.Params.Add("log_level", logLevel)
+	}
+
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+
+	logCh := make(chan string, 64)
+	go func() {
+		defer close(logCh)
+		defer resp.Body.Close()
+		scanner := bufio.NewScanner(resp.Body)
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			if !scanner.Scan() {
+				return
+			}
+
+			// An empty string signals to the caller that
+			// the scan is done, so make sure we only emit
+			// that when the scanner says it's done, not if
+			// we happen to ingest an empty line.
+			if text := scanner.Text(); text != "" {
+				logCh <- text
```

Comment:
It would be good to put this in a `select` statement with a default so that if the buffer is full for some reason we can still cancel the command via canceling the context (and potentially print some kind of error indicating that messages were dropped). If you want to get fancy you could buffer the messages internally and drain when the channel unblocks; another fancy thing would be if you have a default statement and keep track of how many times you hit it, then if the function quits out (or the channel unblocks) you can print a note indicating that X messages were dropped.

Reply:
All of this sounds really good, but I think much of the functionality you describe already exists on the server side, in [helper/monitor/monitor.go](https://github.com/hashicorp/vault/pull/8477/files#diff-fbbd42e134a56cfb61a2a798f248b364).  Is there any way to share that functionality between these two locations?  Is it valuable to be checking for dropped messages on both the client and server side?  I don't mean to second-guess your judgement - mostly I'm just curious if there's a way to avoid duplication of effort/code (or if that's just the nature of this particular beast).

Reply:
This pattern is on the server side too but it's running in two totally different contexts. You don't want to block the server because of this, but you also don't want to block the client because of this. For instance, if someone hits Ctrl-C while logCh is full, you don't want the command to just hang until text comes in, you want it to actually exit.

Reply:
Gotcha, makes sense.  I'll see if there's a way I can abstract out the details of all of this so it can be used in both places.